### DISCUSSION TITLE
Does not detect failed Sudo calls

### DISCUSSION THREAD

**Discussion by stanly-johnson** [polkascan/py-substrate-interface#295] [Q&A] [answered]
The receipt.is_success and receipt.error_message are all good even though the extrinsic failed, the error is within the Sudo module and the process_events() does not parse it. Is it meant to work this way?

Output of receipt.triggered_events
```
[<scale_info::16(value={'phase': 'ApplyExtrinsic', 'extrinsic_idx': 1, 'event': {'event_index': '0508', 'module_id': 'Balances', 'event_id': 'Withdraw', 'attributes': ('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY', 2050039581502301)}, 'event_index': 5, 'module_id': 'Balances', 'event_id': 'Withdraw', 'attributes': ('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY', 2050039581502301), 'topics': []})>, <scale_info::16(value={'phase': 'ApplyExtrinsic', 'extrinsic_idx': 1, 'event': {'event_index': '0200', 'module_id': 'Sudo', 'event_id': 'Sudid', 'attributes': {'Err': 'BadOrigin'}}, 'event_index': 2, 'module_id': 'Sudo', 'event_id': 'Sudid', 'attributes': {'Err': 'BadOrigin'}, 'topics': []})>, <scale_info::16(value={'phase': 'ApplyExtrinsic', 'extrinsic_idx': 1, 'event': {'event_index': '0507', 'module_id': 'Balances', 'event_id': 'Deposit', 'attributes': ('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY', 2050039581502301)}, 'event_index': 5, 'module_id': 'Balances', 'event_id': 'Deposit', 'attributes': ('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY', 2050039581502301), 'topics': []})>, <scale_info::16(value={'phase': 'ApplyExtrinsic', 'extrinsic_idx': 1, 'event': {'event_index': '0000', 'module_id': 'System', 'event_id': 'ExtrinsicSuccess', 'attributes': {'weight': 39547010000, 'class': 'Normal', 'pays_fee': 'Yes'}}, 'event_index': 0, 'module_id': 'System', 'event_id': 'ExtrinsicSuccess', 'attributes': {'weight': 39547010000, 'class': 'Normal', 'pays_fee': 'Yes'}, 'topics': []})>]
```

**Comment by arjanz** [comment]
I don't have more context, but from the events you provided, it seems like the extrinsic succeeded and not failed:
The third event (triggered by extrinsic with index 1 in that block) is an `ExtrinsicSuccess` event.

Where did you see that your extrinsic failed? Did you perhaps saw events that are triggered by another extrinsic in the same block?

**Comment by stanly-johnson** [comment]
Sudo extrinsic is shown as Success but the actual execution failed, you can see this if you scroll along the logs `{'event_index': '0200', 'module_id': 'Sudo', 'event_id': 'Sudid', 'attributes': {'Err': 'BadOrigin'}}`

The process_events() misses it since the extrinsic is marked as Success, I am getting around by looking for it specifically like this
```
for event in receipt.triggered_events:
        if event.value['module_id'] == 'Sudo':
            if 'Err' in str(event.value['attributes']):
```

Again not sure if its meant to work like this, but adding this to the process_events() would be useful.

**Comment by arjanz** [comment, marked as answer]
Ok gotcha, I missed that attribute at first glance. 

I can understand this seems not intuitive, but this is not something that should be altered in the library, because it is just translating the behaviour of this call in the Substrate runtime. 

Apparently in the `Sudo` pallet a `BadOrigin` is interpreted as successful (executed with a valid Sudo account, but not allowed for given call). And for example a `RequireSudo` is a failure, because it is executed with an invalid Sudo account. 

This is a design decision of that Substrate runtime pallet, so I believe best practice is to translate this as such to the applications interfacing with it in the most generic and unmodified way.  Added exceptions in a library will just add confusing I'm afraid, because then behaviour wil differ from other libraries like PolkadotJS.

PolkadotJS is showing the same behaviour:

![Screenshot 2022-05-05 at 17 35 04](https://user-images.githubusercontent.com/5286904/166960897-b8be07ac-17bb-4036-a1b1-3ba1aa97ea36.png)

![Screenshot 2022-05-05 at 17 37 44](https://user-images.githubusercontent.com/5286904/166960944-15bf8de9-b323-4411-a905-4b3b3b05f623.png)

**Comment by stanly-johnson** [comment]
Makes sense. Thanks
